core/src/main/java/org/apache/calcite/rel/rules/MinusToAntiJoinRule.java

```diff
@@ -74,10 +74,17 @@ protected MinusToAntiJoinRule(Config config) {
       RelDataType leftFieldType = left.getRowType().getFieldList().get(i).getType();
       RelDataType rightFieldType = right.getRowType().getFieldList().get(i).getType();
 
-      conditions.add(
-          relBuilder.isNotDistinctFrom(
-              rexBuilder.makeInputRef(leftFieldType, i),
-              rexBuilder.makeInputRef(rightFieldType, i + fieldCount)));
+      if (!leftFieldType.isNullable() && !rightFieldType.isNullable()) {
```

Comment:
I wonder whether this is enough.
You can join on ROW types, and in this case I wonder whether the check should be done on each field.

Reply:
I suddenly thought of a question. We need to judge whether it is `nullable` more precisely, so as to convert `is not distinct` from to `=`?

Reply:
it will work for scalar types as it is; you could check that

Reply:
Will it have any impact if we use is not distinct from uniformly and do not make more complex judgments (return to the logic of the previous version)?

Reply:
the previous is correct, but this one can be slightly more efficient.
However, perhaps this kind of rewrite should be in the scope of a different rule, which could perform it everywhere?

Reply:
So I change it back?

Reply:
Personally I think it's cleaner if each optimization is done by a separate transformation rather than each transformation trying to perform many small optimizations by itself.

Reply:
I agree with you, I will convert the new commit soon.